### DISCUSSION TITLE
Noise gate for suppression of background noises

### DISCUSSION THREAD

**Discussion by JohannesBrx** [jamulussoftware/jamulus#1186] [Ideas]
Especially when jamming with large ensembles or choirs, some channels potentially contribute distracting noises *during the pauses*, ranging from notebook fan noise to car noise from a nearby street. For several of these noises, it is not easy to find a simple remedy.

A [noise gate](https://en.wikipedia.org/wiki/Noise_gate) could attenuate these noises by lowering or muting the channel during pauses (of the singer or musician). Although there are more sophisticated *band-limited* noise gates, which operate on individual frequency bands, I would like to focus here on a simple noise gate which either opens or closes the gate for the whole channel. The assumption is that the background noise only distracts during pauses but is less distracting during singing or jamming.

It would be advantageous to have this audio filter within Jamulus since most of the people using Jamulus might not be able to use a digital audio workstation featuring a noise gate. The noise gate could be implemented in a similar way than the reverb filter in the `CAudioReverb` class. In my test version, I could not notice an increase of latency by using the gate.

If there is general interest, I could prepare a pull request with working code.

(In issue #270, a noise gate was discussed for reducing the *audio data*. However, the intention is completely different here.)

**Edit**: I clarified the application and type of the noise gate.

**Comment by ann0see** [comment]
Great idea! I'd support it (especially for choirs without good microphones). I hope it doesn't change sound too much.

For me the question here would be:

1. Does it add more latency? (As you said probably not)
2. Does it add complexity for developers or users (-> KISS principle)? If yes: can we handle it differently?
2. Can I disable it if I don't want it (Probably yes).

**Reply by JohannesBrx** [reply]
In my current version I placed a checkbox between panning and reverb: 
![image](https://user-images.githubusercontent.com/8805876/110219615-ad39d080-7ec0-11eb-822f-15d9d754350d.png)

**Reply by ann0see** [reply]
This looks great and simple.
Do you already have a code to test? OFC, I'd wait for some other comments.

**Reply by JohannesBrx** [reply]
I have code in a [branch](https://github.com/JohannesBrx/jamulus/tree/NoiseGate) in my own fork. The build is currently running and artifacts will be available [here](https://github.com/JohannesBrx/jamulus/actions/runs/627990477), including other extensions.

**Reply by ann0see** [reply]
Great! I'll test it soon (probably with high gain and a lot of white noise) and afterwards with a very low signal?

**Reply by JohannesBrx** [reply]
Yes, you could try a very sensitive omnidirectional microphone to capture fan noise from your computer or other environment noises. Currently, the threshold for the gate cannot be adjusted and thus very low signals might be suppressed accidentally.

**Reply by ann0see** [reply]
Disclaimer: I'm not an audio engineer!

Tested it and the threshold is maybe a bit high (at least on low noise (= I sang quite quietly) the gate is closed)? Probably it's not something for everybody (especially those who have a quiet environment or people who sing quiet songs).

Edit: Just saw your comment:

> thus very low signals might be suppressed accidentally.

**Comment by gene96817** [comment]
What is a noise gate? Conceptually I think there are two parts, the filter and the suppression logic. First, what is the equivalent filter function?

Since a problem channel is also a problem client, can we assume there is adequate spare computing at the client. We could consider using an existing noise cancellation code. The negative is noise cancellation added delay. If only noisy clients have to pay the penalty, using noise cancellation would be a convenient solution. Things like fan noise would be an easy fix.

**Reply by JohannesBrx** [reply]
I think noise cancellation would be much more complex than a noise gate -- both regarding the computational load and the complexity of implementation. You would probably have to operate in the frequency domain, and thus require the computation of an FFT an inverse FFT on the client.

**Reply by gene96817** [reply]
Noise cancellation is a DSP trick (and not a simple FFT). The significance of the computational load depends on our view of the clients. We could simply not run the DSP if there is insufficient spare client capacity.  DSP would be the most complete approach to noise suppression. This is why I ask, what is the noise gate filter function. The simpler the hack, the lower the computation load and the more incomplete is the concept of noise. So, it really depends on what we believe is an adequate noise model.

**Reply by JohannesBrx** [reply]
I am speaking of a noise gate described, e.g., [here](https://en.wikipedia.org/wiki/Noise_gate), which does not exhibit a special *filter* function. When the gate is close, it simply attenuates the signal by a fixed amount or suppresses it completely.

**Reply by gene96817** [reply]
Thank you for the article. I can understand how rock performers value the noise gate. It is less useful to acoustic styles. It could be very useful for a radio engineer to prevent overdriving the modulator and creating (spectrum) splatter. 

From the article:

> A common application is with electric guitar to remove hum and hiss noise caused by distortion effects units. A noise gate does not remove noise from the signal itself; when the gate is open, both the signal and the noise will pass through. Even though the signal and the unwanted noise are both present in open gate status, the noise is not as noticeable. The noise becomes most noticeable during periods where the main signal is not present, such as a bar of rest in a guitar solo.

This is not a method to remove noise. It is a method to automatically reduce the gain to make the noise less noticeable until there is enough of the main signal that the noise is not as noticeable I don't think our expectations for non-rock styles is on target.

**Reply by JohannesBrx** [reply]
> I can understand how rock performers value the noise gate.

It is just one possible application of the noise gate. 

> This is not a method to remove noise. It is a method to automatically reduce the gain to make the noise less noticeable until there is enough of the main signal that the noise is not as noticeable I don't think our expectations for non-rock styles is on target.

My original intention of the noise gate was driven by the background noise in choir rehearsals during pauses (possibly of only one voice, e.g., the tenor voice) or during instructor explanations. It is not a method to suppress noise during singing. However, I would say that most of the mentioned background noise (like computer fan noise) is hardly noticeable when the person is singing (or jamming).

I clarified the type and application of the noise in my initial contribution of this discussion.

**Reply by ann0see** [reply]
After thinking a bit more about this, I'm a bit concerned about usability:

1. The current implementation doesn't support very low signal -> you'd need to be able to adjust the threshold. But that's not something for inexperienced users. Which this feature is targeted to. 

2. You can't enable it by default since it would confuse people with very low signal/during a quiet part of a piece. "Why don't they hear me?"

Also it's always possible to click "mute myself" during breaks.

**Reply by dcorson-ticino-com** [reply]
I understand the problems leading to this suggestion, but it appears to me to be adding to the concerns for the inexperienced user, they need to turn it on normally, but off for quiet passages.  That is not optimal.

(Personally once a rehearsal has started I never even look at the computer.  If I had to make manipulations during the rehearsal it would bother me greatly.  And I am not an inexperienced user.)

**Reply by gene96817** [reply]
@JohannesBrx    Your objective makes sense.

**Reply by JohannesBrx** [reply]
Regarding low signals being suppressed: I also heard feedback from other testers that the noise gate settings are too *aggressive*. But I am quite confident that I will find lower thresholds which will work in most cases. To find out acceptable values, I probably will have to implement a slider for adjustment first.

I will let you know when the settings are optimized.

**Comment by dcorson-ticino-com** [comment]
Do I see this correctly that the client who is noisy needs to turn on his noise gate?

**Reply by JohannesBrx** [reply]
Yes, that's right. Applying the noise gate on the mixed signal would not give satisfying results.

**Comment by ghost** [comment]
Noise gates, Background noise cancelling/microphones, are best used ouside of Jamulus.
I would not be opposed to removing the Reverb functionality from Jamulus also.

**Reply by JohannesBrx** [reply]
I would estimate that most of the people using Jamulus would not be able to setup a digital audio workstation with a reverb filter or noise gate. Even for the more experienced users it is cumbersome to use BlackHole or other internal loopback devices to forward the audio to a separate program. Therefore, in my opinion, Jamulus *should* provide basic audio filters.

**Reply by dakhubgit** [reply]
That sounds like reinventing the wheel a lot.  Shouldn't Jamulus rather support some plugin interface like LADSPA or LAV or LV2 or VST?  That would open the door to a lot of readily available functionality.

**Reply by JohannesBrx** [reply]
Like you would expect a text editor to have a basic spell checking I would expect Jamulus to have basic audio filters like a noise gate "onboard" without further installation. Implementing a plugin interface is probably much more complicated than implementing a few basic audio filters. 

Of course, if we would integrate many audio filters, this would be a reinvention of the wheel.

**Reply by hoffie** [reply]
I think this should be decided case by case. I agree that Jamulus should try to focus on the core features.

Personally, I'd like to have certain features which make Jamulus sounding better without any external complexity. This makes Jamulus accessible to more people. Noise is a relevant problem for larger groups such as choirs. I can't tell yet if the approach taken here solves the problem and does so in a way which could be included into the official Jamulus release.

At the same time we have to avoid getting each and every possible feature into Jamulus as all of this adds maintenance cost, might introduce bugs or performance penalties (just a generic statement, not targeted at this feature specifically).

**Comment by Hk1020** [comment]
What I have noticed is that noise increases the more clients join. Every client seems to contribute a little bit of noise even with very good audio interfaces and that adds up resulting in the annoying background noise. I am talking about 10 to 20 clients here. Once that drops to 5 or so it is relaxingly quiet. It is hard to identify a single client which is the main contributor. So I am wondering if we are addressing the right problem here.

**Reply by JohannesBrx** [reply]
I would say that it is a second or very similar use case to reduce the overall noise. The assumption is, however, that the summed up noise is hardly noticeable when the whole choir or band is active. The noise gate will only help to reduce noise in the pauses.

**Comment by JohannesBrx** [comment]
For those who are interested: I have updated my current [branch](https://github.com/JohannesBrx/jamulus/tree/NoiseGate). It is now possible to set the (open) threshold of the noise gate between -60...-30dB by using a slider. (The closing threshold is -10dB lower.) The slider is just there for experimentation.
